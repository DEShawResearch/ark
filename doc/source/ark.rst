
The Ark class
-------------

.. autoclass:: ark.Ark
    :members:
    :show-inheritance:


