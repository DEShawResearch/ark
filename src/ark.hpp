#ifndef __ark_ark_hpp
#define __ark_ark_hpp

#include "base.hpp"
#include "printer.hpp"
#include "parser.hpp"
#include "exception.hpp"
#include "argv.hpp"
#include "reader.hpp"

#endif
