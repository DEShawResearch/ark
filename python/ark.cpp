#include <pybind11/pybind11.h>
#include <sstream>
#include <set>
#include <ark/ark.hpp>

using namespace pybind11;
using namespace Ark;

namespace {
    struct _ {
        _() {
            module mod;
            // collections.Iterable is deprecated in 3.7 in favor of
            // collections.abc, but the latter is unavailable in 2.7.
            try {
                mod = module::import("collections.abc");
            } catch (error_already_set&) {
                mod = module::import("collections");
            }
            iterable = mod.attr("Iterable");
        };
        handle iterable;
    } _init;
    typedef std::set<PyObject*> ObjSet;
    void make_object(ark& a, handle obj, ObjSet circular=ObjSet()) {
        if (circular.count(obj.ptr())) {
            throw std::runtime_error("Circular reference detected");
        }
        circular.insert(obj.ptr());
        handle File = module::import("ark").attr("File");
        if (isinstance<str>(obj)) {
            a.be(Atom).atom() = std::string(str(obj));
        } else if (isinstance<float_>(obj)) {
            /* I'm using repr(obj) here because Python's repr generates
             * a compact, but non-lossy, string representation of floats.
             * printf("%.17g"), would also be non-lossy but generates 
             * something like "0.10000000000000001" for 0.1, while Python's
             * repr returns "0.1". */
            a.be(Atom).atom() = std::string(repr(obj));
        } else if (isinstance<dict>(obj)) {
            table_t& t = a.be(Table).table();
            for (auto pair : reinterpret_borrow<dict>(obj)) {
                ark& v = t[std::string(str(pair.first))];
                make_object(v, pair.second, circular);
            }
        } else if (isinstance(obj, _init.iterable)) {
            vector_t& v = a.be(Vector).vector();
            for (auto elem : obj) {
                v.resize(v.size()+1);
                make_object(v.back(), elem, circular);
            }
        } else if (isinstance(obj, File)) {
            // stick a funny character at the front which will be
            // handled by the ark printer
            std::string s;
            s.push_back(0xff);
            s.append(std::string(str(obj.attr("text"))));
            a.be(Atom).atom() = s;
        } else if (obj.ptr()==Py_None) {
            a.be(None);
        } else if (obj.ptr()==Py_True) {
            a.be(Atom).atom() = "true";
        } else if (obj.ptr()==Py_False) {
            a.be(Atom).atom() = "false";
        } else {
            a.be(Atom).atom() = std::string(str(obj));
        }
    }

    object to_object(ark const& a, bool convert_strings) {
        switch (a.kind()) {
            case Atom:
                if (convert_strings) {
                    const char* s = a.atom().c_str();
                    /* empty string? */
                    if (*s=='\0') return str(s);
                    /* true/false? */
                    if (!strcasecmp(s, "true")) return bool_(true);
                    if (!strcasecmp(s, "false")) return bool_(false);
                    /* integer? */
                    char* end = NULL;
                    long i = strtol(s, &end, 0);
                    if (*end=='\0') return int_(i);
                    /* float? */
                    end = NULL;
                    double d = strtod(s, &end);
                    if (*end=='\0') return float_(d);
                    /* plain old string */
                    return str(s);
                }
                return str(a.atom().c_str());
            case Vector:
                {
                    list L;
                    for (ark const& e : a.vector()) {
                        L.append(to_object(e,convert_strings));
                    }
                    return L;
                }
                break;
            case Table:
                {
                    dict D;
                    for (table_t::value_type const& t : a.table()) {
                        D[str(t.first.c_str())] = to_object(t.second,convert_strings);
                    }
                    return D;
                }
                break;
            case None: ;
        };
        return none();
    }
}

PYBIND11_MODULE(_ark, m) {

    enum_<kind_t>(m, "kind")
        .value("None", None) 
        .value("Atom", Atom)
        .value("Vector", Vector)
        .value("Table", Table)
        .export_values();

    class_<ark>(m, "ark")
        .def(init<>())
        ;

    class_<parser>(m, "parser")
        .def(init<>())
        .def("parse_string",
             (ark&(parser::*)(ark&, std::string const&) const) &parser::parse)
        .def("parse_keyvals",
             (ark&(parser::*)(ark&, std::string const&) const) &parser::parse_keyvals)
        .def("parse_file", &parser::parse_file)
        ;

    class_<printer::printer_ref>(m, "printer_ref")
        .def("output", [](const printer::printer_ref& p) {
            std::stringstream ss;
            p.output(ss);
            return ss.str();
            })
        ;

    class_<printer>(m, "printer")
        .def(init<>())
        .def("width",   &printer::width)
        .def("indent",  &printer::indent)
        .def("whitespace",&printer::whitespace)
        .def("no_delim",&printer::no_delim)
        .def("open_tables",&printer::open_tables)
        .def("flatten", &printer::flatten)
        .def("__call__", &printer::operator())
        ;

    m.def("from_object", [](object obj) {
            ark a; make_object(a, obj); return a;
            });
    m.def("to_object", to_object);

    m.def("valid_key", Ark::key_t::valid_key);
}

