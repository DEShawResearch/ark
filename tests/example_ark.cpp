#include <ark/ark.hpp>

using namespace Ark;

int main(int argc,char **argv) {

  ark a,b;

  a = "hey";

  a.be(Vector).vector().push_back("hello");
  a.be(Vector).vector().push_back("world");

  a.be(Table);
  a.table()["global_cell"].be(Table);
  a.table()["global_cell"].table()["partition"].be(Vector);
  a.table()["global_cell"].table()["partition"].vector().push_back("1");
  a.table()["global_cell"].table()["partition"].vector().push_back("3");
  a.table()["global_cell"].table()["partition"].vector().push_back("9");
  a.table()["global_cell"].table()["partition"].vector()[2] = "7";

  fdump(stdout,a);
  printf("\n");

  fdump(stdout, parse("{ x=[1 2 3] z=? y={z = hey w='string string'}}"));
  printf("\n");

  a.table()["sigma"] = "1.234";
  a.table()["topology"] = "january";

  b.be(Vector);
  b.vector().push_back("10");
  b.vector().push_back("41");
  b.vector().push_back("21");

  a.table()["global_cell"].be(Table).table()["partition"].swap(b);

  fdump(stdout,a);
  printf("\n");
  fdump(stdout,b);
  printf("\n");
  
  const ark *t;
  if ((t=a.get("global_cell"))) {
    fdump(stdout,*t);
    printf("\n");
    if ((t=t->get("partition"))) {
      fdump(stdout,*t);
      printf("\n");
    }
  }
  if ((t=a.get("global_cell")->get("partition"))) {
    fdump(stdout,*t);
    printf("\n");
  }
  const atom_t *c;
  if ((c=a.get("global_cell")->get("partition")->get(1)->get_atom()))
    printf("%s\n",c->c_str());
}
