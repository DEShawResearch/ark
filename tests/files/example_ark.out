{global_cell={partition=["1""3""7"]}}
{x=["1""2""3"]y={w="string string"z="hey"}z=?}
{global_cell={partition=["10""41""21"]}sigma="1.234"topology="january"}
["1""3""7"]
{partition=["10""41""21"]}
["10""41""21"]
["10""41""21"]
41
